codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -158,67 +152,44 @@ public void incrementWindowSize(ChannelHandlerContext ctx, Http2Stream stream, i
             // Update the stream window and write any pending frames for the stream.
             FlowState state = state(stream);
             state.incrementStreamWindow(delta);
-            frameSent = false;
+            needFlush = false;
             state.writeBytes(state.writableWindow());
-            if (frameSent) {
+            if (needFlush) {
                 flush();
             }
         }
     }
 
     @Override
-    public ChannelFuture sendFlowControlledFrame(ChannelHandlerContext ctx, Http2Stream stream,
-            ByteBuf data, int padding, boolean endStream, ChannelPromise promise) {
+    public ChannelFuture sendFlowControlledData(ChannelHandlerContext ctx, Http2Stream stream,
+            FlowControlledPayload payload, ChannelPromise promise) {
+        if (stream.state() == Http2Stream.State.IDLE) {
+          throw new IllegalStateException("Cannot send DATA on IDLE stream");
+        }
         checkNotNull(ctx, "ctx");
         checkNotNull(promise, "promise");
-        checkNotNull(data, "data");
+        checkNotNull(payload, "data");
         if (this.ctx != null && this.ctx != ctx) {
             throw new IllegalArgumentException("Writing data from multiple ChannelHandlerContexts is not supported");
         }
-        if (padding < 0) {
-            throw new IllegalArgumentException("padding must be >= 0");
-        }
-
         // Save the context. We'll use this later when we write pending bytes.
         this.ctx = ctx;
-
         try {
             FlowState state = state(stream);
-
-            int window = state.writableWindow();
-            boolean framesAlreadyQueued = state.hasFrame();
-
-            FlowState.Frame frame = state.newFrame(promise, data, padding, endStream);
-            if (!framesAlreadyQueued && window >= frame.size()) {
-                // Window size is large enough to send entire data frame
-                frame.write();
-                ctx.flush();
-                return promise;
-            }
-
-            // Enqueue the frame to be written when the window size permits.
+            FlowState.Frame frame = state.newFrame(payload);
             frame.enqueue();
-
-            if (framesAlreadyQueued || window <= 0) {
-                // Stream already has frames pending or is stalled, don't send anything now.
-                return promise;
+            needFlush = false;
+            state.writeBytes(state.writableWindow());
+            if (needFlush) {
+              ctx.flush();
+              needFlush = false;
```

Comment:
We may be able to get away with not having to explicitly set `needFlush = false` before every write too.
```java
flush() {
  if (needFlush) {
    ctx.flush();
    needFlush = false;
  }
}
```